src/main/java/com/puppycrawl/tools/checkstyle/xpath/SuppressionsFileGenerator.java

```diff
@@ -0,0 +1,176 @@
+            "ImportControl", "JavadocVariable", "LeftCurly", "MethodParamPad",
+            "MultipleVariableDeclarations", "NestedForDepth", "NestedIfDepth", "NestedTryDepth",
+            "NPathComplexity", "OneStatementPerLine", "OuterTypeNumber", "RequireThis", "RightCurly"
+    };
+
+    /**
+     * Since currently not all checks support xpath suppressions, this method returns true if check
+     * supports xpath and false otherwise.
+     * @param checkName the name of the check.
+     * @return returns true if name is valid
+     */
+    private boolean isValidCheckName(String checkName) {
+        boolean result = false;
+        for (String supportedCheck : supportedChecks) {
+            if (checkName.contains(supportedCheck)) {
+                result = true;
+                break;
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Creates a new {@code SuppressionFileGenerator} instance.
+     * Sets the output to a defined stream.
+     * @param outputStream the stream to write logs to.
+     * @param tabWidth distance between tab stop position
+     */
+    public SuppressionsFileGenerator(OutputStream outputStream,
+                                     int tabWidth) {
+        this.tabWidth = tabWidth;
+        writer = new PrintWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
+
+        writer.println("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
+        writer.println("<!DOCTYPE suppressions PUBLIC");
+        writer.println("    \"-//Puppy Crawl//DTD Suppressions Xpath Experimental 1.2//EN\"");
+        writer.println("    \"http://checkstyle.sourceforge.net/dtds"
+                + "/suppressions_1_2_xpath_experimental.dtd\">");
+    }
+
+    @Override
+    public void auditStarted(AuditEvent event) {
+        writer.println("<suppressions>");
+    }
+
+    @Override
+    public void auditFinished(AuditEvent event) {
+        writer.println("</suppressions>");
+        writer.flush();
+    }
+
+    @Override
+    public void fileStarted(AuditEvent event) {
+
+    }
+
+    @Override
+    public void fileFinished(AuditEvent event) {
+
+    }
+
+    @Override
+    public void addError(AuditEvent event) {
+        final String checkName = event.getSourceName();
+
+        if (isValidCheckName(checkName)) {
+            final LocalizedMessage message = event.getLocalizedMessage();
+            final int lineNumber = message.getLineNo();
+            final int columnNumber = message.getColumnNo();
+            final int tokenType = message.getTokenType();
+            final String fileName = event.getFileName();
+
+            final File file = new File(fileName);
+            try {
+                final FileText fileText = new FileText(file,
+                        StandardCharsets.UTF_8.name());
+                final DetailAST rootAst = JavaParser.parseFile(file,
+                        JavaParser.Options.WITH_COMMENTS);
```

Comment:
Because it is an audit listener and don't have the AST, we have to re-parse the Java file again.
There isn't even any check to see if file was already reparsed in case the same file generated multiple violations.

Reply:
what is another way to solve this issue, what can be used instead of audit listener?

Reply:
@timurt , please review https://github.com/checkstyle/checkstyle/pull/5883#discussion_r192606843